models/perm/access/repo_permission.go

```diff
@@ -143,135 +143,162 @@ func (p *Permission) ColorFormat(s fmt.State) {
+		if err != nil {
+			return
+		}
 
-	// Collaborators on organization
-	if is {
-		for _, u := range repo.Units {
-			perm.UnitsMode[u.Type] = perm.AccessMode
+		// Org stuff
+		if !repo.Owner.IsOrganization() {
+			continue
 		}
-	}
 
-	// get units mode from teams
-	teams, err := organization.GetUserRepoTeams(ctx, repo.OwnerID, user.ID, repo.ID)
-	if err != nil {
-		return
-	}
+		perms[repo.ID].UnitsMode = make(map[unit.Type]perm_model.AccessMode)
 
-	// if user in an owner team
-	for _, team := range teams {
-		if team.AccessMode >= perm_model.AccessModeAdmin {
-			perm.AccessMode = perm_model.AccessModeOwner
-			perm.UnitsMode = nil
+		// Collaborators on organization
+		if is {
+			for _, u := range repo.Units {
+				perms[repo.ID].UnitsMode[u.Type] = perms[repo.ID].AccessMode
+			}
+		}
+
+		// TODO: load teams in one query
+		// get units mode from teams
+		var teams organization.TeamList
+		teams, err = organization.GetUserRepoTeams(ctx, repo.OwnerID, user.ID, repo.ID)
+		if err != nil {
 			return
 		}
-	}
 
-	for _, u := range repo.Units {
-		var found bool
+		// if user in an owner team
 		for _, team := range teams {
-			teamMode := team.UnitAccessMode(ctx, u.Type)
-			if teamMode > perm_model.AccessModeNone {
-				m := perm.UnitsMode[u.Type]
-				if m < teamMode {
-					perm.UnitsMode[u.Type] = teamMode
-				}
-				found = true
+			if team.AccessMode >= perm_model.AccessModeAdmin {
+				perms[repo.ID].AccessMode = perm_model.AccessModeOwner
+				perms[repo.ID].UnitsMode = nil
+				continue repoLoop
 			}
 		}
 
-		// for a public repo on an organization, a non-restricted user has read permission on non-team defined units.
-		if !found && !repo.IsPrivate && !user.IsRestricted {
-			if _, ok := perm.UnitsMode[u.Type]; !ok {
-				perm.UnitsMode[u.Type] = perm_model.AccessModeRead
+		for _, u := range repo.Units {
+			var found bool
+			for _, team := range teams {
+				teamMode := team.UnitAccessMode(ctx, u.Type)
+				if teamMode > perm_model.AccessModeNone {
+					m := perms[repo.ID].UnitsMode[u.Type]
+					if m < teamMode {
+						perms[repo.ID].UnitsMode[u.Type] = teamMode
+					}
+					found = true
+				}
+			}
+
+			// for a public repo on an organization, a non-restricted user has read permission on non-team defined units.
+			if !found && !repo.IsPrivate && !user.IsRestricted {
+				if _, ok := perms[repo.ID].UnitsMode[u.Type]; !ok {
+					perms[repo.ID].UnitsMode[u.Type] = perm_model.AccessModeRead
+				}
 			}
 		}
-	}
 
-	// remove no permission units
-	perm.Units = make([]*repo_model.RepoUnit, 0, len(repo.Units))
-	for t := range perm.UnitsMode {
-		for _, u := range repo.Units {
-			if u.Type == t {
-				perm.Units = append(perm.Units, u)
+		// remove no permission units
+		perms[repo.ID].Units = make([]*repo_model.RepoUnit, 0, len(repo.Units))
+		for t := range perms[repo.ID].UnitsMode {
+			for _, u := range repo.Units {
+				if u.Type == t {
+					perms[repo.ID].Units = append(perms[repo.ID].Units, u)
```

Comment:
Can't we add a `break` here to improve the performance even more?
Or can units be added twice?
```suggestion
		// remove no permission units
		perms[repo.ID].Units = make([]*repo_model.RepoUnit, 0, len(repo.Units))
		for t := range perms[repo.ID].UnitsMode {
			for _, u := range repo.Units {
				if u.Type == t {
					perms[repo.ID].Units = append(perms[repo.ID].Units, u)
					break
```